backend/btrixcloud/subs.py

```diff
@@ -0,0 +1,236 @@
+        """create org for new subscription"""
+        subscription = Subscription(
+            subId=create.subId, status=create.status, planId=create.planId
+        )
+
+        new_org = await self.org_ops.create_org(
+            quotas=create.quotas, subscription=subscription
+        )
+
+        result = {"added": True, "id": new_org.id}
+
+        if create.firstAdminInviteEmail:
+            is_new, token = await self.org_ops.invites.invite_user(
+                InviteToOrgRequest(
+                    email=create.firstAdminInviteEmail, role=UserRole.OWNER
+                ),
+                user,
+                self.user_manager,
+                org=new_org,
+                headers=dict(request.headers),
+            )
+            if is_new:
+                result["invited"] = "new_user"
+            else:
+                result["invited"] = "existing_user"
+            result["token"] = token
+
+        await self.add_sub_event(create)
+
+        return result
+
+    async def update_subscription(self, update: SubscriptionUpdate) -> dict[str, bool]:
+        """update subs"""
+
+        org = await self.org_ops.update_subscription_data(update)
+
+        if not org:
+            raise HTTPException(
+                status_code=404, detail="org_for_subscription_not_found"
+            )
+
+        await self.add_sub_event(update)
+        return {"updated": True}
+
+    async def cancel_subscription(self, cancel: SubscriptionCancel) -> dict[str, bool]:
+        """delete subscription data, and if readOnlyOnCancel is true, the entire org"""
+
+        org = await self.org_ops.cancel_subscription_data(cancel)
+
+        if not org:
+            raise HTTPException(
+                status_code=404, detail="org_for_subscription_not_found"
+            )
+
+        # extra sanity check, shouldn't ever be true
+        if not org.subscription or org.subscription.subId != cancel.subId:
+            return {"canceled": False, "deleted": False}
+
+        # mark as read-only even if deleting, in case deletion
+        # takes some time
+        deleted = False
+
+        await self.org_ops.update_read_only(
+            org, readOnly=True, readOnlyReason="subscriptionCanceled"
+        )
+
+        if not org.subscription.readOnlyOnCancel:
+            await self.org_ops.delete_org_and_data(org, self.user_manager)
+            deleted = True
+
+        await self.add_sub_event(cancel)
+        return {"canceled": True, "deleted": deleted}
+
+    async def add_sub_event(
+        self, event: Union[SubscriptionCreate, SubscriptionUpdate, SubscriptionCancel]
+    ) -> None:
+        """add a subscription event to the db"""
+        data = event.dict(exclude_unset=True)
+        data["type"] = event.type
+        data["timestamp"] = datetime.utcnow()
+        await self.subs.insert_one(data)
+
+    async def list_sub_events(self):
+        """list subscription events"""
+        # note: should add sorting/filtering eventually
```

Comment:
maybe worth implementing this and pagination now? I'm happy to jump in and do that if you'd like :)

Reply:
Done! :)